### Importing Libraries

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('../data/raw/sales_data.csv')

In [3]:
df.shape

(302010, 30)

In [4]:
df.columns

Index(['Transaction_ID', 'Customer_ID', 'Name', 'Email', 'Phone', 'Address',
       'City', 'State', 'Zipcode', 'Country', 'Age', 'Gender', 'Income',
       'Customer_Segment', 'Date', 'Year', 'Month', 'Time', 'Total_Purchases',
       'Amount', 'Total_Amount', 'Product_Category', 'Product_Brand',
       'Product_Type', 'Feedback', 'Shipping_Method', 'Payment_Method',
       'Order_Status', 'Ratings', 'products'],
      dtype='str')

In [5]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 302010 entries, 0 to 302009
Data columns (total 30 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Transaction_ID    301677 non-null  float64
 1   Customer_ID       301702 non-null  float64
 2   Name              301628 non-null  str    
 3   Email             301663 non-null  str    
 4   Phone             301648 non-null  float64
 5   Address           301695 non-null  str    
 6   City              301762 non-null  str    
 7   State             301729 non-null  str    
 8   Zipcode           301670 non-null  float64
 9   Country           301739 non-null  str    
 10  Age               301837 non-null  float64
 11  Gender            301693 non-null  str    
 12  Income            301720 non-null  str    
 13  Customer_Segment  301795 non-null  str    
 14  Date              301651 non-null  str    
 15  Year              301660 non-null  float64
 16  Month             301737 non-nu

## Understanding Data

- This is a retail sales data of year 2023 and 2024, which contains wide range of information about sales, transactions, products and customers all combined in a single csv file.
- Total number of records in the dataset is 302010 and number of columns are 30.

In [6]:
df.head(10)

,Transaction_ID,Customer_ID,Name,Email,Phone,Address,City,State,Zipcode,Country,...,Total_Amount,Product_Category,Product_Brand,Product_Type,Feedback,Shipping_Method,Payment_Method,Order_Status,Ratings,products
0,8691788.0,37249.0,Michelle Harrington,Ebony39@gmail.com,1.414787e+09,3959 Amanda Burgs,Dortmund,Berlin,77985.0,Germany,...,324.086270,Clothing,Nike,Shorts,Excellent,Same-Day,Debit Card,Shipped,5.0,Cycling shorts
1,2174773.0,69749.0,Kelsey Hill,Mark36@gmail.com,6.852900e+09,82072 Dawn Centers,Nottingham,England,99071.0,UK,...,806.707815,Electronics,Samsung,Tablet,Excellent,Standard,Credit Card,Processing,4.0,Lenovo Tab
2,6679610.0,30192.0,Scott Jensen,Shane85@gmail.com,8.362160e+09,4133 Young Canyon,Geelong,New South Wales,75929.0,Australia,...,1063.432799,Books,Penguin Books,Children's,Average,Same-Day,Credit Card,Processing,2.0,Sports equipment
3,7232460.0,62101.0,Joseph Miller,Mary34@gmail.com,2.776752e+09,8148 Thomas Creek Suite 100,Edmonton,Ontario,88420.0,Canada,...,2466.854021,Home Decor,Home Depot,Tools,Excellent,Standard,PayPal,Processing,4.0,Utility knife
4,4983775.0,27901.0,Debra Coleman,Charles30@gmail.com,9.098268e+09,5813 Lori Ports Suite 269,Bristol,England,48704.0,UK,...,248.553049,Grocery,Nestle,Chocolate,Bad,Standard,Cash,Shipped,1.0,Chocolate cookies
5,6095326.0,41289.0,Ryan Johnson,Haley12@gmail.com,3.292677e+09,532 Ashley Crest Suite 014,Brisbane,New South Wales,74430.0,Australia,...,1185.167224,Electronics,Apple,Tablet,Good,Express,PayPal,Pending,4.0,Lenovo Tab
6,5434096.0,97285.0,Erin Lewis,Arthur76@gmail.com,1.578355e+09,600 Brian Prairie Suite 497,Kitchener,Ontario,47545.0,Canada,...,630.115295,Electronics,Samsung,Television,Bad,Standard,Cash,Processing,1.0,QLED TV
7,2344675.0,26603.0,Angela Fields,Tanya94@gmail.com,3.668096e+09,237 Young Curve,Munich,Berlin,86862.0,Germany,...,46.588070,Clothing,Zara,Shirt,Bad,Same-Day,Cash,Processing,1.0,Dress shirt
8,4155845.0,80175.0,Diane Clark,Martin39@gmail.com,6.219780e+09,8823 Mariah Heights Apt. 263,Wollongong,New South Wales,39820.0,Australia,...,2630.714413,Grocery,Nestle,Chocolate,Bad,Same-Day,Cash,Delivered,1.0,Dark chocolate
9,4926148.0,31878.0,Lori Bell,Jessica33@gmail.com,6.004895e+09,6225 William Lodge,Cologne,Berlin,64317.0,Germany,...,3976.112295,Home Decor,Home Depot,Decorations,Excellent,Standard,Cash,Delivered,4.0,Candles


In [7]:
df.tail(10)

,Transaction_ID,Customer_ID,Name,Email,Phone,Address,City,State,Zipcode,Country,...,Total_Amount,Product_Category,Product_Brand,Product_Type,Feedback,Shipping_Method,Payment_Method,Order_Status,Ratings,products
302000,9862022.0,99656.0,Amber Fields,Donald38@gmail.com,6.472397e+09,50022 Antonio Valley Suite 498,Plymouth,England,92925.0,UK,...,151.805498,Grocery,Nestle,Chocolate,Bad,Express,Cash,Pending,1.0,Chocolate mousse
302001,5561211.0,35974.0,Jordan Hall,Holly86@gmail.com,2.062575e+09,806 Alan Flats,Ballarat,New South Wales,99633.0,Australia,...,2401.466964,Home Decor,Bed Bath & Beyond,Bathroom,Bad,Same-Day,Cash,Processing,1.0,Soap dispenser
302002,8961631.0,79479.0,Jason Welch,Jason36@gmail.com,6.279294e+09,764 Garcia Flat,Hamilton,Ontario,61218.0,Canada,...,2659.976987,Home Decor,Home Depot,Tools,Excellent,Express,Cash,Pending,5.0,Level
302003,2844206.0,18799.0,Angel Hood,Joseph24@gmail.com,2.825445e+09,7593 Joseph Trace Suite 382,Cairns,New South Wales,39837.0,Australia,...,2384.717299,Electronics,Apple,Tablet,Average,Same-Day,Cash,Pending,2.0,Amazon Fire Tablet
302004,4833982.0,94117.0,Kara Hart,Tammy37@gmail.com,7.108672e+09,872 Robinson Harbors Apt. 328,Charlotte,Missouri,65301.0,USA,...,2362.120301,Clothing,Nike,Shorts,Excellent,Standard,Cash,Delivered,4.0,Chino shorts
302005,4246475.0,12104.0,Meagan Ellis,Courtney60@gmail.com,7.466354e+09,389 Todd Path Apt. 159,Townsville,New South Wales,4567.0,Australia,...,973.962984,Books,Penguin Books,Fiction,Bad,Same-Day,Cash,Processing,1.0,Historical fiction
302006,1197603.0,69772.0,Mathew Beck,Jennifer71@gmail.com,5.754305e+09,52809 Mark Forges,Hanover,Berlin,16852.0,Germany,...,285.137301,Electronics,Apple,Laptop,Excellent,Same-Day,Cash,Processing,5.0,LG Gram
302007,7743242.0,28449.0,Daniel Lee,Christopher100@gmail.com,9.382530e+09,407 Aaron Crossing Suite 495,Brighton,England,88038.0,UK,...,182.105285,Clothing,Adidas,Jacket,Average,Express,Cash,Shipped,2.0,Parka
302008,9301950.0,45477.0,Patrick Wilson,Rebecca65@gmail.com,9.373222e+09,3204 Baird Port,Halifax,Ontario,67608.0,Canada,...,120.834784,Home Decor,IKEA,Furniture,Good,Standard,Cash,Shipped,4.0,TV stand
302009,2882826.0,53626.0,Dustin Merritt,William14@gmail.com,9.518927e+09,143 Amanda Crescent,Tucson,West Virginia,25242.0,USA,...,2382.233417,Home Decor,Home Depot,Decorations,Average,Same-Day,Cash,Shipped,2.0,Clocks


In [8]:
df.columns = df.columns.str.strip().str.lower()
df.columns

Index(['transaction_id', 'customer_id', 'name', 'email', 'phone', 'address',
       'city', 'state', 'zipcode', 'country', 'age', 'gender', 'income',
       'customer_segment', 'date', 'year', 'month', 'time', 'total_purchases',
       'amount', 'total_amount', 'product_category', 'product_brand',
       'product_type', 'feedback', 'shipping_method', 'payment_method',
       'order_status', 'ratings', 'products'],
      dtype='str')

In [9]:
df.isnull().sum()

transaction_id      333
customer_id         308
name                382
email               347
phone               362
address             315
city                248
state               281
zipcode             340
country             271
age                 173
gender              317
income              290
customer_segment    215
date                359
year                350
month               273
time                350
total_purchases     361
amount              357
total_amount        350
product_category    283
product_brand       281
product_type          0
feedback            184
shipping_method     337
payment_method      297
order_status        235
ratings             184
products              0
dtype: int64

In [10]:
df.nunique()

transaction_id      294461
customer_id          86766
name                159390
email                52897
phone               299295
address             299329
city                   130
state                   54
zipcode              93978
country                  5
age                     53
gender                   2
income                   3
customer_segment         3
date                   366
year                     2
month                   12
time                 83677
total_purchases         10
amount              299297
total_amount        299305
product_category         5
product_brand           18
product_type            33
feedback                 4
shipping_method          3
payment_method           4
order_status             4
ratings                  5
products               318
dtype: int64

In [32]:
df.describe(include='all')

,transaction_id,customer_id,name,email,phone,address,city,state,zipcode,country,...,total_amount,product_category,product_brand,product_type,feedback,shipping_method,payment_method,order_status,ratings,products
count,3.016770e+05,301702.000000,301628,301663,3.016480e+05,301695,301762,301729,301670.000000,301739,...,301660.000000,301727,301729,302010,301826,301673,301713,301775,301826.000000,302010
unique,NaN,NaN,159390,52897,NaN,299329,130,54,NaN,5,...,NaN,5,18,33,4,3,4,4,NaN,318
top,NaN,NaN,Michael Smith,Michael59@gmail.com,NaN,11313 Steven Place Apt. 141,Chicago,England,NaN,USA,...,NaN,Electronics,Pepsi,Water,Excellent,Same-Day,Credit Card,Delivered,NaN,Spring water
freq,NaN,NaN,144,93,NaN,2,21652,63059,NaN,95223,...,NaN,71196,30292,24459,100755,104156,90116,130449,NaN,2508
mean,5.495823e+06,55006.553934,NaN,NaN,5.501464e+09,NaN,NaN,NaN,50298.951019,NaN,...,1367.651156,NaN,NaN,NaN,NaN,NaN,NaN,NaN,3.162670,NaN
std,2.595565e+06,26005.675200,NaN,NaN,2.596017e+09,NaN,NaN,NaN,28972.807134,NaN,...,1128.998515,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1.320827,NaN
min,1.000007e+06,10000.000000,NaN,NaN,1.000049e+09,NaN,NaN,NaN,501.000000,NaN,...,10.003750,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1.000000,NaN
25%,3.247930e+06,32469.250000,NaN,NaN,3.255061e+09,NaN,NaN,NaN,25425.000000,NaN,...,438.724278,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2.000000,NaN
50%,5.499657e+06,55012.000000,NaN,NaN,5.505812e+09,NaN,NaN,NaN,50602.500000,NaN,...,1041.117547,NaN,NaN,NaN,NaN,NaN,NaN,NaN,3.000000,NaN
75%,7.739509e+06,77511.000000,NaN,NaN,7.749860e+09,NaN,NaN,NaN,75252.000000,NaN,...,2029.999853,NaN,NaN,NaN,NaN,NaN,NaN,NaN,4.000000,NaN


### Key Points

- No similarities found between name and email id.
- 4 duplicate records exist.
- out of 302010 records, 294461 record's transactions are unique and non null.
- No of unique customer_id : 86766
- No of unique name of customers : 159390.

***Customer Information***
- `'customer_id'` : It is a unique identifier of customer data, no two different customer can have same `customer_id`. 
- `'name'` : It represents name of the person placing order, it might be possible that single `customer_id` linked with more than one name `name`.
- `'email'` : It represents email id of the customer or the person making order.
- `'phone'` : It is the contact number of the person. it can also be classified as unique identifier.
- `'age'` : It is the age of the user. widely varies from 18 years to 70 years.
- `'gender'` : customers genders are male and female only.
- `'income'` : customers are grouped into categories based on their income as low, medium, and high.
- `'address'` : it represents delivery addresses of the users.
- `'city'` : city of user placing order for.
- `'state'` : state of user placing order for.
- `'zipcode'` : zipcode of location user placing order for.
- `'country'` : country of user placing order in. 
- `'customer_segment'` : customers categorised as regular, premium and new.

***Transaction Details***
- `'transaction_id'` : it represents transactions and its unique identifier. around 333 records shows NA in transaction_id. 
- `'total_purchases'` : it is the quantity of items ordering of. better to rename as quantity.
- `'total_amount'` : it is the total amount, price x quantity. 
- `'feedback'` : customer review on the product. 
- `'shipping_method'` : same day shipping, one day delivery, express delivery. 
- `'payment_method'` : debit, credit, cash.
- `'order_status'` : pending, shipped, processing, delivered.

***Product Information***
- `'products'` : name of product.
- `'amount'` : amount represents price of the product.
- `'product_category'` : category of product.
- `'product_brand'` : brand of product.
- `'product_type'` : type of product.
- `'ratings'` : how much product liked or disliked by any customer.

***Temporal Information***
- `'year'` : year of order placed
- `'month'` : month of order placed
- `'date'` : date of order placed, better to transform it into day of month.
- `'time'` : time of order placed.

### Problems Found 
- Transaction_id and customer_id dtype need to change in integer from float.
- remove duplicate records.

***Potential Analysis***
- Customer segmentation analysis based on demographics, purchase behavior, and feedback.
- Sales trend analysis over time to identify peak seasons or trends.
- Product performance analysis to determine popular categories, brands, or types.
- Geographic analysis to understand regional preferences and trends.
- Payment and shipping method analysis to optimize services.
- Customer satisfaction analysis based on feedback and order status.

### Data Preprocessing

In [ ]:
# This lead to 16.2% memory optimization

dtype_mapping = {
   'transaction_id': 'Int32',
    'customer_id': 'Int32',
    'age' : 'Int8',
    'total_purchases': 'Int8',
    'year': 'Int16',
    'amount': 'float16',
    'total_amount': 'float32',
    'ratings' : 'Int8',
}

for col, dtype in dtype_mapping.items():
  if col in df.columns:
    df[col] = df[col].astype(dtype)


  # df['age'] = (
  #     pd.to_numeric(df['age'], errors='coerce')
  #     .round()
  #     .astype('Int8')
  #   )

In [ ]:
# # 1. Downcasting columns for memory optimization
# dtype_mapping = {
  
#     # Categoricals (Reduces memory from ~8 bytes/pointer + string to 1-2 bytes/row)
#     # 'city': 'category',
#     # 'state': 'category',
#     # 'country': 'category',
#     # 'gender': 'category',
#     # 'customer_segment': 'category',
#     # 'product_category': 'category',
#     # 'product_brand': 'category',
#     # 'product_type': 'category',
#     # 'shipping_method': 'category',
#     # 'payment_method': 'category',
#     # 'order_status': 'category',

    
#     # Downcasted Numerics
#      # IDs as nullable integers or strings
#     'transaction_id': 'Int32',
#     'customer_id': 'Int32',
#     'age' : 'Int8',
#     'total_purchases': 'Int8',
#     'year': 'Int16',
#     'month' : 'Int8',
#     'time' : 'datetime64[ns]',
#     'ratings': 'float32',
#     'amount': 'float16',
#     'total_amount': 'float32',
#     'ratings' : 'Int4',
# }


# def optimize_retail_dtypes(df: pd.DataFrame) -> pd.DataFrame:
#   # Apply basic mappings

#   # Combine date and time into a single standard datetime index
#   # if 'date' in df.columns and 'time' in df.columns:
#   #   df['timestamp'] = pd.to_datetime(
#   #       df['date'].astype(str) + ' ' + df['time'].astype(str), errors='coerce'
#   #   )
#   #   df = df.drop(columns=['date', 'time', 'year', 'month'])

#   # Drop non-predictive PII / unstructured text to maximize memory savings
#   # pii_cols = [
#   #     'name',
#   #     'email',
#   #     'phone',
#   #     'address',
#   #     'zipcode',
#   #     'feedback',
#   #     'products',
#   # ]
#   # df = df.drop(columns=[col for col in pii_cols if col in df.columns])

#   # return df

In [44]:
# Removing Duplicates based on all columns
df = df.drop_duplicates()
df = df.dropna(subset=['transaction_id'])

In [ ]:
df[df.duplicated(subset={'transaction_id', 'customer_id','phone','date','zipcode'}, keep=False)].sort_values(by = 'transaction_id')

,transaction_id,customer_id,name,email,phone,address,city,state,zipcode,country,...,total_amount,product_category,product_brand,product_type,feedback,shipping_method,payment_method,order_status,ratings,products
165944,1005039,84239,John Tate,Andrea27@gmail.com,9.064051e+09,7563 Pittman Tunnel,Bonn,Berlin,35327.0,Germany,...,1114.320068,Grocery,Coca-Cola,Water,Bad,Same-Day,Debit Card,Pending,1,Flavored water
301783,1005039,84239,John Tate,Andrea27@gmail.com,9.064051e+09,7563 Pittman Tunnel,Bonn,Berlin,35327.0,Germany,...,1114.320068,Grocery,Coca-Cola,Water,Bad,Same-Day,Cash,Pending,1,Sparkling water
273886,1010919,12847,Tyler Ward,Louis68@gmail.com,4.553740e+09,945 Justin Street,Vancouver,Ontario,79443.0,Canada,...,1361.597046,Electronics,Samsung,Television,Good,Same-Day,Credit Card,Delivered,3,LED TV
300651,1010919,12847,Tyler Ward,Louis68@gmail.com,4.553740e+09,945 Justin Street,Vancouver,Ontario,79443.0,Canada,...,1361.597046,Electronics,Samsung,Television,Good,Same-Day,Cash,Delivered,3,LED TV
97360,1019366,21198,Timothy Montgomery,Cynthia36@gmail.com,8.622075e+09,385 Gabrielle Estate,Winnipeg,Ontario,69618.0,Canada,...,1417.142456,Grocery,Nestle,Snacks,Average,Express,PayPal,Shipped,2,Pretzels
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
5022,9979219,93801,Jeffrey Lewis,Amy89@gmail.com,4.922819e+09,577 Kramer Tunnel,Portsmouth,England,94941.0,UK,...,113.231758,Home Decor,IKEA,Lighting,Average,Standard,PayPal,Pending,2,Desk lamps
130903,9986359,24076,Michael Hawkins,Ricky21@gmail.com,4.303454e+09,487 James Trail,Montreal,Ontario,17890.0,Canada,...,1392.184326,Home Decor,IKEA,Furniture,Good,Standard,Debit Card,Delivered,3,Dresser
301019,9986359,24076,Michael Hawkins,Ricky21@gmail.com,4.303454e+09,487 James Trail,Montreal,Ontario,17890.0,Canada,...,1392.184326,Home Decor,IKEA,Furniture,Good,Standard,Cash,Pending,3,TV stand
263943,9992663,10935,Travis Lawrence,Thomas93@gmail.com,6.406085e+09,26767 York Spurs,Quebec City,Ontario,97494.0,Canada,...,1783.312988,Electronics,Mitsubhisi,Mitsubishi 1.5 Ton 3 Star Split AC,Good,Same-Day,Credit Card,Delivered,4,Ductless AC


In [73]:
df

,transaction_id,customer_id,name,email,phone,address,city,state,zipcode,country,age,gender,income,customer_segment,date,year,month,time,total_purchases,amount,total_amount,product_category,product_brand,product_type,feedback,shipping_method,payment_method,order_status,ratings,products
0,8691788.0,37249.0,Michelle Harrington,Ebony39@gmail.com,1.414787e+09,3959 Amanda Burgs,Dortmund,Berlin,77985.0,Germany,21.0,Male,Low,Regular,9/18/2023,2023.0,September,22:03:55,3.0,108.028757,324.086270,Clothing,Nike,Shorts,Excellent,Same-Day,Debit Card,Shipped,5.0,Cycling shorts
1,2174773.0,69749.0,Kelsey Hill,Mark36@gmail.com,6.852900e+09,82072 Dawn Centers,Nottingham,England,99071.0,UK,19.0,Female,Low,Premium,12/31/2023,2023.0,December,8:42:04,2.0,403.353907,806.707815,Electronics,Samsung,Tablet,Excellent,Standard,Credit Card,Processing,4.0,Lenovo Tab
2,6679610.0,30192.0,Scott Jensen,Shane85@gmail.com,8.362160e+09,4133 Young Canyon,Geelong,New South Wales,75929.0,Australia,48.0,Male,Low,Regular,4/26/2023,2023.0,April,4:06:29,3.0,354.477600,1063.432799,Books,Penguin Books,Children's,Average,Same-Day,Credit Card,Processing,2.0,Sports equipment
3,7232460.0,62101.0,Joseph Miller,Mary34@gmail.com,2.776752e+09,8148 Thomas Creek Suite 100,Edmonton,Ontario,88420.0,Canada,56.0,Male,High,Premium,5/8/2023,2023.0,May,14:55:17,7.0,352.407717,2466.854021,Home Decor,Home Depot,Tools,Excellent,Standard,PayPal,Processing,4.0,Utility knife
4,4983775.0,27901.0,Debra Coleman,Charles30@gmail.com,9.098268e+09,5813 Lori Ports Suite 269,Bristol,England,48704.0,UK,22.0,Male,Low,Premium,1/10/2024,2024.0,January,16:54:07,2.0,124.276524,248.553049,Grocery,Nestle,Chocolate,Bad,Standard,Cash,Shipped,1.0,Chocolate cookies
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
302005,4246475.0,12104.0,Meagan Ellis,Courtney60@gmail.com,7.466354e+09,389 Todd Path Apt. 159,Townsville,New South Wales,4567.0,Australia,31.0,Male,Medium,Regular,1/20/2024,2024.0,January,23:40:29,5.0,194.792597,973.962984,Books,Penguin Books,Fiction,Bad,Same-Day,Cash,Processing,1.0,Historical fiction
302006,1197603.0,69772.0,Mathew Beck,Jennifer71@gmail.com,5.754305e+09,52809 Mark Forges,Hanover,Berlin,16852.0,Germany,35.0,Female,Low,New,12/28/2023,2023.0,December,2:55:45,1.0,285.137301,285.137301,Electronics,Apple,Laptop,Excellent,Same-Day,Cash,Processing,5.0,LG Gram
302007,7743242.0,28449.0,Daniel Lee,Christopher100@gmail.com,9.382530e+09,407 Aaron Crossing Suite 495,Brighton,England,88038.0,UK,41.0,Male,Low,Premium,2/27/2024,2024.0,February,2:43:49,3.0,60.701761,182.105285,Clothing,Adidas,Jacket,Average,Express,Cash,Shipped,2.0,Parka
302008,9301950.0,45477.0,Patrick Wilson,Rebecca65@gmail.com,9.373222e+09,3204 Baird Port,Halifax,Ontario,67608.0,Canada,41.0,Male,Medium,New,9/3/2023,2023.0,September,11:20:31,1.0,120.834784,120.834784,Home Decor,IKEA,Furniture,Good,Standard,Cash,Shipped,4.0,TV stand


In [ ]:
# 1. Define explicit dtype mappings for load/transformation
dtype_mapping = {
    # Categoricals (Reduces memory from ~8 bytes/pointer + string to 1-2 bytes/row)
    'city': 'category',
    'state': 'category',
    'country': 'category',
    'gender': 'category',
    'customer_segment': 'category',
    'product_category': 'category',
    'product_brand': 'category',
    'product_type': 'category',
    'shipping_method': 'category',
    'payment_method': 'category',
    'order_status': 'category',
    # Downcasted Numerics
    'total_purchases': 'Int16',
    'year': 'Int16',
    'ratings': 'float32',
    'amount': 'float32',
    'total_amount': 'float32',
    # IDs as nullable integers or strings
    'transaction_id': 'Int32',
    'customer_id': 'Int32',
}


def optimize_retail_dtypes(df: pd.DataFrame) -> pd.DataFrame:
  # Apply basic mappings
  for col, dtype in dtype_mapping.items():
    if col in df.columns:
      df[col] = df[col].astype(dtype)

  # Clean and parse age safely to Int8
  if 'age' in df.columns:
    df['age'] = (
        pd.to_numeric(df['age'], errors='coerce')
        .round()
        .astype('Int8')
    )

  # Combine date and time into a single standard datetime index
  if 'date' in df.columns and 'time' in df.columns:
    df['timestamp'] = pd.to_datetime(
        df['date'].astype(str) + ' ' + df['time'].astype(str), errors='coerce'
    )
    df = df.drop(columns=['date', 'time', 'year', 'month'])

  # Drop non-predictive PII / unstructured text to maximize memory savings
  pii_cols = [
      'name',
      'email',
      'phone',
      'address',
      'zipcode',
      'feedback',
      'products',
  ]
  df = df.drop(columns=[col for col in pii_cols if col in df.columns])

  return df